### PR TITLE
Refactor miner: Isolate EVM processing

### PR DESCRIPTION
**Pull request by prasannavl**
## Summary

- Cleanup and isolate the EVM process handling in the miner.

## Implications

- Storage
  - [ ] Database reindex required
  - [ ] Database reindex optional
  - [ ] Database reindex not required
  - [x] None

- Consensus
  - [ ] Network upgrade required
  - [ ] Includes backward compatible changes
  - [ ] Includes consensus workarounds
  - [ ] Includes consensus refactors
  - [x] None

### CODE REVIEW

File: src/miner.cpp [hunk 1]
```diff
@@ -649,6 +709,93 @@ void BlockAssembler::SortForBlock(const CTxMemPool::setEntries& package, std::ve
...
+            checkedDfTxHashSet.erase(addrTxs[addrKey.nonce]->GetTx().GetHash());
+            addrTxs[addrKey.nonce] = txIter;
+            auto count{addrKey.nonce};
+            for (const auto& [nonce, entry] : addrTxs) {
+                checkedDfTxHashSet.erase(entry->GetTx().GetHash());
+                replaceByFee.emplace(count, entry);
+                ++count;
+            }
+            evmAddressTxsMap.erase(addrKey.address);
+            evm_remove_txs_by_sender(evmQueueId, addrKey.address);
```
Comment (sieniven) [inline src/miner.cpp:778]: Same thing as caught in PR #2220. When we call evm_remove_txs_by_sender, the current pipeline is such that we are actually removing all transactions from that specific sender in the txqueue. This actually means that if for example we send 5 evm txs from 0 to 5 nonces, and then an evm tx of nonce 3 with higher prepay gas is submitted which replaces the evm tx nonce 3 - we are wiping all txs from 0 to 5 here. This test isnt actually covered in feature_evm, since the nonce replacement is nicely replacing nonce 0 in the test case which misses this specific edge case.

File: src/miner.cpp [hunk 2]
```diff
@@ -649,6 +709,93 @@ void BlockAssembler::SortForBlock(const CTxMemPool::setEntries& package, std::ve
...
+                    ++it;
+                }
+            }
+            // Buggy code to fix below:
+            checkedDfTxHashSet.erase(addrTxs[addrKey.nonce]->GetTx().GetHash());
+            addrTxs[addrKey.nonce] = txIter;
+            auto count{addrKey.nonce};
+            for (const auto& [nonce, entry] : addrTxs) {
+                checkedDfTxHashSet.erase(entry->GetTx().GetHash());
+                replaceByFee.emplace(count, entry);
```
Comment (sieniven) [inline src/miner.cpp:774]: Similar to PR #2220 - is there a reason to be using count instead of directly using the nonce matching the evm tx when storing descendant evm txs inside this index?

File: src/miner.cpp (hunk 1 above)
Comment (sieniven) [reply src/miner.cpp:778]: Resolving as the current pipeline is correct. We are wiping all of our indexes whenever we replace the evm tx (both ancestors and descendants), and adding them into the replacebyfee index.

File: src/miner.cpp (hunk 2 above)
Comment (Bushstar) [reply src/miner.cpp:774]: We did not have the nonce before and they should have been in nonce order. Will update this.

File: src/miner.cpp [hunk 3]
```diff
@@ -69,6 +69,40 @@ struct XVM {
...
+    uint64_t nonce;
+
+    bool operator<(const EvmAddressWithNonce& item) const {
+        return std::tie(address, nonce) < std::tie(item.address, item.nonce);
+    }
+};
+
+struct EvmPackageContext {
+    // Used to track EVM TXs by sender and nonce.
+    std::map<EvmAddressWithNonce, uint64_t> feeMap;
```
Comment (Jouzo) [inline src/miner.cpp:85]:
```suggestion
    // Used to track EVM TX fee by sender and nonce.
    std::map<EvmAddressWithNonce, uint64_t> feeMap;
```

**Review by Jouzo — DISMISSED** [review]

**Review by Jouzo — DISMISSED** [review]

**Review by Bushstar — APPROVED** [review]

**Review by Jouzo — APPROVED** [review]
